client/verta/verta/runtime.py

```diff
@@ -0,0 +1,173 @@
+    """
+    Sets the thread-local validate flag boolean value.
+    """
+    _THREAD.validate = flag
+    return _THREAD.validate
+
+
+def validate_json(value: Any) -> str:
+    """
+    Attempt to serialize the provided value to JSON.
+    Raise an error on failure.
+    """
+    try:
+        return json.dumps(value)
+    except json.JSONDecodeError as json_err:
+        raise Exception("Unable to convert logging data to JSON") from json_err
+    except TypeError as type_err:
+        raise Exception("Unable to serialize log value to JSON.") from type_err
+
+
+def log(key: str, value: Any) -> None:
+    """
+    Updates current logging context dict with provided key and value.
+    For use within a model's ``predict()`` method to collect logging
+    context, which is stored in thread-local variables.
+
+    Parameters
+    ----------
+    key : str
+        String value to use as data label (key) in logging context.
+    value : Any
+        Any JSON serializable value you wish to include in the context logs.
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    .. code-block:: python
+       :emphasize-lines: 10,13
+
+        from verta import runtime
+
+        # Sample model code:
+        class MyModel(VertaModelBase):
+            def __init__(self, artifacts):
+                "ok"
+
+            def predict(self, data):
+                logs = {'some_stuff': 'I_care_about'}
+                runtime.log('my_logging_key', logs)
+                prediction_output = 'some_output'
+                more_logs = ['a', 'list', 'of', 'things']
+                runtime.log('things', more_logs)
+                return prediction_output
+
+    """
+    if _get_validate_flag():
+        validate_json(value)
+    local_context: Dict[str, Any] = _get_thread_logs()
+    local_context.update({key: value})
+    _set_thread_logs(local_context)
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Uses thread-local variables to store context in a thread-safe manner.
+
+    Parameters
+    ----------
+    validate : Optional[bool]
+        If true, each individual call to ``runtime.log('key', value)`` will
+        verify that the value provided is JSON serializable.
+
+    Attributes
+    ----------
+    logs : Dict[str, Any]
+        Dictionary of the current logging context.  If called after exiting
+        the context manager, the final complete log entry is returned.
+    """
+    def __init__(self, validate: Optional[bool] = False):
+        self.validate = validate
+        self.logs_dict = dict()
+
+    def __enter__(self):
+        """
+        Ensure an empty logging context to start and set validation flag.
+        """
+        _set_thread_logs(dict())
+        if self.validate:
+            _set_validate_flag(True)
+        return self
+
+    def __exit__(self, *args):
+        """
+        Capture the final complete log entry in a class variable, ensure an
```

Comment:
```suggestion
        Capture the final complete log entry in an instance variable, ensure an
```

Technically [an instance variable rather than a class variable](https://docs.python.org/3/tutorial/classes.html#class-and-instance-variables) 😄

Reply:
Fixed